java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java

```diff
@@ -285,30 +336,110 @@ private API getXPAPI(ExpressionTree et) {
+
+      // when argumentIndex is specified, if mit's argument at argIndex doesn't match,
+      // skip to next method property map
+      if (argumentIndexSpecified) {
+        int argIndex = ((Long) currentMethodProperties.get(ARGUMENT_INDEX_STRING)).intValue();
+        if (argIndex < mit.getArguments().size()) {
+          ExpressionTree argTree = mit.getArguments().get(argIndex);
+          Symbol argSym = ASTHelpers.getSymbol(argTree);
+          if (!isArgumentMatchesFlagName(argTree, argSym)) {
+            continue;
           }
+        } else {
+          continue;
         }
       }
+
+      // check both return and receiver if both are specified
+      // else check only the one which is specified
+      if (returnSpecified && receiverSpecified) {
+        if (checkMethodTreeMatchesMethodProperty(
+                mit, RETURN_TYPE_STRING, (String) currentMethodProperties.get(RETURN_TYPE_STRING))
+            && checkMethodTreeMatchesMethodProperty(
+                mit,
+                RECEIVER_TYPE_STRING,
+                (String) currentMethodProperties.get(RECEIVER_TYPE_STRING))) {
+          return matchAPIToFlagTypeString((String) currentMethodProperties.get(FLAG_TYPE_KEY));
+        }
+      } else if (returnSpecified || receiverSpecified) {
+        if (returnSpecified) {
+          if (checkMethodTreeMatchesMethodProperty(
+              mit, RETURN_TYPE_STRING, (String) currentMethodProperties.get(RETURN_TYPE_STRING))) {
+            return matchAPIToFlagTypeString((String) currentMethodProperties.get(FLAG_TYPE_KEY));
+          }
+        } else if (receiverSpecified) {
+          if (checkMethodTreeMatchesMethodProperty(
+              mit,
+              RECEIVER_TYPE_STRING,
+              (String) currentMethodProperties.get(RECEIVER_TYPE_STRING))) {
+            return matchAPIToFlagTypeString((String) currentMethodProperties.get(FLAG_TYPE_KEY));
+          }
+        }
+      } else {
+        return matchAPIToFlagTypeString((String) currentMethodProperties.get(FLAG_TYPE_KEY));
+      }
     }
     return API.UNKNOWN;
   }
 
+  private API matchAPIToFlagTypeString(String flagType) {
+    if (TREATED_KEY.equalsIgnoreCase(flagType)) {
+      return API.IS_TREATED;
+    } else if (CONTROL_KEY.equalsIgnoreCase(flagType)) {
+      return API.IS_CONTROL;
+    } else if (EMPTY_KEY.equalsIgnoreCase(flagType)) {
+      return API.DELETE_METHOD;
+    } else if (TREATMENT_GROUP_KEY.equalsIgnoreCase(flagType)) {
+      return API.IS_TREATMENT_GROUP_CHECK;
+    } else {
+      return API.UNKNOWN;
+    }
+  }
+
+  private boolean checkMethodTreeMatchesMethodProperty(
+      MethodInvocationTree mit, String propKey, String propValue) {
+    MemberSelectTree mst = ((MemberSelectTree) mit.getMethodSelect());
+    if (RETURN_TYPE_STRING.equals(propKey)) {
+      String mReturn = ASTHelpers.getReturnType(mst).toString();
+      return Pattern.compile(propValue, Pattern.CASE_INSENSITIVE).matcher(mReturn).find();
```

Comment:
Good catch. But I'm not really sure how much we can pre-compute it.

**Why this case is rare (enough)**
It is triggered only when `returnType` or `receiverType` is specified - for method names that are specified in properties.json

The `for` loop with currentMethodProperties is only triggered once in the vast majority of cases. With the `continue` refactoring, `checkMethodTreeMatchesMethodProperty` is only called once if either return or receiver are specified and twice if both are.

**Why it's difficult to easily speedup**
The `mReturn/mReceive` can be obtained only when mit/mst is passed in. That means only the `Pattern.compile(propValue, Pattern.CASE_INSENSITIVE)` can be pre-computed, but even this depends on the value given for `returnType` or `receiverType`, which may or may not be unique for a given config file. We might get some speedup by precomputing all `Pattern.compile("boolean", Pattern.CASE_INSENSITIVE)` for example.

To do that, we would have to iterate over the config file after parsing it and compute all `Pattern.compile` if return or receiver is specified. (These would then be stored in a Map<String, Pattern.Compile> somewhere?)

But I think this may not provide that much of a speedup and seems more confusing code-wise.

Let me know if I've missed something or if you disagree.

Reply:
Thanks for the detailed response. This just got me thinking that we really don't need a pattern for either returnType or receiverType. These can be exact fully qualified names that can be specified in the json and the comparison can be with that. The flexibility of having a pattern makes the implementation (and documentation) complex but doesn't yield any additional benefits given the user has to specify methodName and argumentIndex precisely. Can we get rid of the pattern for these types and simplify the code, and update the documentation for this?

Reply:
Sounds good. I think there may be some use to having pattern match for receiver, but its a rare enough that it may not be worth the complexity.